boskos/janitor/janitor_test.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"k8s.io/test-infra/boskos/common"
+)
+
+type fakeBoskos struct {
+	lock      sync.Mutex
+	wg        sync.WaitGroup
+	resources []common.Resource
+}
+
+// Create a fake client
+func CreateFakeBoskos(resources int) *fakeBoskos {
+	fb := &fakeBoskos{}
+	for i := 0; i < resources; i++ {
+		fb.resources = append(fb.resources,
+			common.Resource{
+				Name:  fmt.Sprintf("res-%d", i),
+				Type:  "project",
+				State: "dirty",
+			})
+	}
+
+	return fb
+}
+
+func (fb *fakeBoskos) Acquire(rtype string, state string, dest string) (string, error) {
+	fb.lock.Lock()
+	defer fb.lock.Unlock()
+
+	for idx := range fb.resources {
+		r := &fb.resources[idx]
+		if r.State == state {
+			r.State = dest
+			fb.wg.Add(1)
+			return r.Name, nil
+		}
+	}
+
+	return "", nil
+}
+
+func (fb *fakeBoskos) ReleaseOne(name string, dest string) error {
+	fb.lock.Lock()
+	defer fb.lock.Unlock()
+
+	for idx := range fb.resources {
+		r := &fb.resources[idx]
+		if r.Name == name {
+			r.State = dest
+			fb.wg.Done()
+			return nil
+		}
+	}
+
+	return fmt.Errorf("No resource %v", name)
+}
+
+// waitTimeout waits for the waitgroup for the specified max timeout.
+// Returns true if waiting timed out.
+func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally
+	case <-time.After(timeout):
+		return true // timed out
+	}
+}
+
+func TestDrain(t *testing.T) {
```

Comment:
Updated and this is just the regular sanity test. Also scaled up to 1000 to see if anything would bust.